contracts/contracts/Multisig.cairo

```diff
@@ -525,4 +571,27 @@ func _set_transaction_calldata{
         calldata=calldata + 1,
     )
     return ()
-end
\ No newline at end of file
+end
+
+func _clean_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_index : felt, owners_len : felt
+):
+    if owners_index == owners_len:
+        return ()
+    end
+
+    let (owner_address) = _owners.read(owners_index)
+    _is_owner.write(owner_address, FALSE)
+    _owners.write(owners_index, 0)
+
+    return _clean_owners(owners_index + 1, owners_len)
+end
+
+func _set_confirmations_required_impl{
+    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
+}(confirmations_required : felt):
+    _confirmations_required.write(confirmations_required)
```

Comment:
I feel the owner&confirmation number check should be here instead of in two separate places. Now it's both in constructor and in `set_confirmations_required`. And what makes it a bit more confusing is that they both now give a different error message.

Reply:
Agreed!

Reply:
I wouldn't agree with this. The point is to preserve invariants and to make sure any workflow makes the same checks, raises the same errors, and preserves all invariants (no surprises).

This is hard to do and follow if this kind of logic is spread all over the contract. Also introduces inconsistencies, like we currently have with error messages for this specific workflow.

Additionally, `_set_owners_impl` (similar fn) currently also checks owners are unique.

So in conclusion, it would be better if the `_set_confirmations_required` could include that range safety check (less code, stronger invariants, fewer bugs).

Reply:
> The point is to preserve invariants

Invariant is that values that are passed to setter is valid. Same is happening in `_set_owners` we've checked values before now we just set them.

> workflow makes the same checks, raises the same errors, and preserves all invariants (no surprises)

Its happening. Some inner function assume that values passed to them are correct, were checked and that's invariant. Business logic functions raise errors and so on.

> This is hard to do and follow if this kind of logic is spread all over the contract. 

Well my opinion is that if setter or getter does smth other than set or get that is break of Single Responsibility Principle. 

> Additionally, _set_owners_impl (similar fn) currently also checks owners are unique.

Similar function in that case would be `_set_owners`. `_set_owners_impl` does pre-checks and business logic and after that passes values to setter which as invariant assumes that values are valid

What I could suggest as compromise here is adding function smth like `verify_and_set_confirmations_required` which makes checks with `owners_len` and then calls `_set_confirmations_required`

Reply:
I'm perfectly fine with adding the `verify_and_set_confirmations_required` 👍 

But, ideas like the Single Responsibility Principle and similar don't really map to blockchain contracts 1:1, because the environment is constrained. The focus is on security and cost (coupled with performance). You have limited and expensive resources, so best engineering practices have evolved to adapt to the environment (cost, tx limits, contract size limits, gas limits, etc.). It's similar to writing firmware code for constrained hardware devices. Sometimes there isn't enough space to be able to structure all the objects and functions as desired, so a tighter integration is required.